zpool import progress kstat

ofaaland opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
When an import requires a long MMP activity check, or when the user
requests pool recovery, the import make take a long time.  The user may
not know why, or be able to tell whether the import is progressing or is
hung.

### Description
<!--- Describe your changes in detail -->

Add a kstat which lists all imports currently being processed by the
kernel (currently only one at a time is possible, but the kstat allows
for more than one).  The kstat is at
/proc/spl/kstat/zfs/import_progress.

The kstat contents are as follows:
```
pool_guid            load_state multihost_secs  max_txg pool_name
16667015954387398856 3           15             0       tank3
```

load_state: the value of spa_load_state
multihost_secs: number of seconds left in the multihost activity check, if any.
max_txg: current spa_load_max_txg, if rewind is occurring

This could be used by outside tools, such as a pacemaker resource agent,
to report import progress, or as a part of manual troubleshooting.  The
zpool import subcommand could also be modified to report this
information.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
Manual testing to confirm contents, including imports both with and without rewind and multihost activity check, together and separately.  Multiple imports issued in userspace at the same time.  Ran zloop and MMP tests in ZTS.

Updated zpool import to use this procfile (See #8646).

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+ */
+
+typedef struct spa_import_progress {
+	uint64_t		pool_guid;	/* unique # for updates */
```
nit: /#/id/

ofaaland replied:
done

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+} spa_import_progress_t;
+
+spa_history_list_t *spa_import_progress_list = NULL;
+unsigned int spa_import_progress_list_size = 16;
```
Why is it important to limit the number of entries to 16?  I understand this has no effect today since concurrent pool imports are not allowed, but I wonder if we want a limit at all.

ofaaland replied:
Since we remove entries from the list whenever an import ends, the list will not stay large, even if it grows large, so we do not need a limit.  And the semantic that there is one and only one entry in the procfile for each in-progress import is easy to describe.  

So I think you are right we do not want a limit.  I removed it.

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+
+	if (size == 0)
+		ASSERT(list_is_empty(&shl->procfs_list.pl_list));
+
```
nit: extra new line

ofaaland replied:
Fixed

ofaaland commented:
Hi @utopiabound is it possible for you to review this patch?  Thanks.

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	}
+
+	if (size == 0)
+		ASSERT(list_is_empty(&shl->procfs_list.pl_list));
```
You can use `IMPLY(size == 0, list_is_empty(&shl->procfs_list.pl_list));` for this check.  Or just remove it.

ofaaland replied:
done.

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	    &spa_import_progress_list->procfs_list,
+	    spa_import_progress_show,
+	    spa_import_progress_show_header,
+	    spa_import_progress_clear,
```
Do we want to support clearing the kstat via writing to proc?  This seems like it could cause confusion.

ofaaland replied:
You're right.  Removed.

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+/*
+ * For update methods, it is important that we start search at the tail of the
+ * list where new records are inserted, so this is normally an O(1) operation.
+ */
```
I'm not sure this comment is true since we're matching on the `pool_guid`.  If multiple long running concurrent imports are allowed the one we're interested in could appear anywhere in the list.

ofaaland replied:
Agreed.

behlendorf commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	if (shl->size == 0)
+		return (0);
+
+
```
nit: extra new line

ofaaland replied:
fixed

behlendorf commented on module/zfs/spa.c:
```diff
@@ -4346,6 +4359,7 @@ spa_load_impl(spa_t *spa, spa_import_type_t type, char **ereport)
 		spa_config_exit(spa, SCL_CONFIG, FTAG);
 	}
 
+	spa_import_progress_remove(spa_guid(spa));
```
Is there value in not removing the kstat entry until the pool is exported?

ofaaland replied:
We discussed this offline, and decided not to keep entries after the import was finished.

The logic was based on what data one might get from the entry after the fact:
* successful import
** MMP check time - this may be different next time, since it is based on I/O latency.
** txg rewound to during recovery - this should be retained permanently, probably via zpool history
** total import time - the time visible from this procfile is just the tryimport or just the real import, not both, and in both cases not including the userspace time.  One could get total time, likely what the user wants, by recording it in userspace.
* failed import
** MMP check time - one could just start the import again, read the time, and then cancel.  This is likely what the admin would think to do anyway.
** txg rewound to which failed - this is an edge case, and something that the admin would be doing by hand.  So if they want to know this, they will likely be monitoring the procfile during the import anyway.

ofaaland commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	    &poolname);
+	if (poolname == NULL)
+		poolname = spa_name(spa);
+	sip->pool_name = strdup(poolname);
```
done

ofaaland commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	    sip = list_prev(&shl->procfs_list.pl_list, sip)) {
+		if (sip->pool_guid == pool_guid) {
+			if (sip->pool_name)
+				strfree(sip->pool_name);
```
done

ofaaland commented on module/zfs/spa_misc.c:
```diff
@@ -2019,6 +2019,236 @@ spa_dirty_data(spa_t *spa)
+	while (shl->size > size) {
+		sip = list_remove_head(&shl->procfs_list.pl_list);
+		if (sip->pool_name)
+			strfree(sip->pool_name);
```
done

ofaaland commented:
@behlendorf , I've addressed your comments above.  I also addressed the issues we discussed offline:
Instead of reporting the start and end times of the activity check, the kstat now reports seconds remaining.  I believe the failure code paths are now properly handled, and the load state is now updated everywhere.

ofaaland commented:
> Instead of reporting the start and end times of the activity check, the kstat now reports seconds remaining

A note for future me: the reasoning is that the kernel knows with certainty how much wall clock time is left in the test.  It uses a monotonic clock source.  However translating that monotonic end time to system time is error-prone due to the possibility that something (e.g. NTP) changes the system time during the test.  So we just give seconds remaining.
